File: src/core/token.ts
import { Position } from "../error/position.ts";
import { TYPES } from "../utils/constants.ts";

export class Token {
  public positionStart: Position | null = null;
  public positionEnd: Position | null = null;
  public constructor(
    public type: TYPES,
    public value: unknown = null,
    positionStart: Position | null = null,
    positionEnd: Position | null = null,
  ) {
    if (positionStart) {
      this.positionStart = positionStart.clone();
      this.positionEnd = positionStart.clone();
      this.positionEnd.advance("");
    }

    if (positionEnd) {
      this.positionEnd = this.positionEnd;
    }
  }

  public match(type: TYPES, value: unknown): boolean {
    return this.type === type && this.value === value;
  }

  public represent(): string {
    if (this.value) return `${this.type}: ${this.value}`;
    return `${this.type}`;
  }
}

File: src/core/token.rs
use crate::utils::{constants::DynType, constants::Tokens, position::Position};
use std::fmt::{Display, Error, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub r#type: Tokens,
    pub value: DynType,
    pub pos_start: Position,
    pub pos_end: Position,
}

impl Display for Token {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        return if self.value == DynType::None {
            write!(f, "[{:?}]", self.r#type)
        } else {
            write!(f, "[{:?}]", self.value)
        };
    }
}

impl Token {
    pub fn new(r#type: Tokens, pos_start: Position, pos_end: Position, value: DynType) -> Token {
        Token {
            r#type,
            value,
            pos_start,
            pos_end,
        }
    }

    pub fn clone(&mut self) -> Token {
        Token {
            r#type: *&self.r#type,
            value: self.value.clone(),
            pos_start: self.pos_start.clone(),
            pos_end: self.pos_end.clone(),
        }
    }

    pub fn matches(self, r#type: Tokens, value: DynType) -> bool {
        return self.r#type == r#type && self.value == value;
    }
}
